Docs basic example cpp sdk (#7835)
ydb-platform/ydb: ydb/docs/en/core/dev/example-app/_includes/example-cpp.md

{% include [steps/02_create_table.md](steps/02_create_table.md) %}

```c++

>     //! Creates sample tables with ExecuteQuery Query Service

Reviewer: ```suggestion
    //! Creates sample tables with the ExecuteQuery method
```

        )", path.c_str());

        return session.ExecuteQuery(query,
            TTxControl::BeginTx(TTxSettings::SerializableRW()).CommitTx()).GetValueSync();
    }));
}
```

{% include [pragmatablepathprefix.md](auxilary/pragmatablepathprefix.md) %}

{% include [steps/04_query_processing.md](steps/04_query_processing.md) %}

To execute YQL queries, use the `ExecuteQuery` method.
The SDK lets you explicitly control the execution of transactions and configure the transaction execution mode using the `TTxControl` class.

In the code snippet below, the transaction is started with the `TTxControl::BeginTx` method. With `TTxSettings`, set the `SerializableRW` transaction execution mode. When all the queries in the transaction are completed, the transaction is automatically completed by explicitly setting `CommitTx()`. The `query` described using the YQL syntax is passed to the `ExecuteQuery` method for execution.

```c++
void SelectSimple(TQueryClient client, const TString& path) {
    TMaybe<TResultSet> resultSet;
    ThrowOnError(client.RetryQuerySync([&path, &resultSet](TSession session) {
        auto query = Sprintf(R"(
            PRAGMA TablePathPrefix("%s");

>             SELECT series_id, title, CAST(CAST(release_date AS Date) AS String) AS release_date

Reviewer 2: release_date column has Uint64 type

>             SELECT series_id, title, CAST(CAST(release_date AS Date) AS String) AS release_date

Reviewer: > release_date column has Uint64 type

That's the reason for the first cast to `Date`, but I was asking about the second one to `String`.

            WHERE series_id IN $series
            ORDER BY season_id;
        )");

        auto paramsBuilder = TParamsBuilder();
        auto& listParams = paramsBuilder
                                    .AddParam("$series")
                                    .BeginList();
        
        for (auto x : {1, 10}) {
            listParams.AddListItem().Uint64(x);
        }
                
        auto parameters = listParams
                                .EndList()
                                .Build()
                                .Build();

        // Executes stream query
        auto resultStreamQuery = client.StreamExecuteQuery(query, TTxControl::NoTx(), parameters).GetValueSync();

        if (!resultStreamQuery.IsSuccess()) {
            return resultStreamQuery;
        }

        // Iterates over results
        bool eos = false;

        while (!eos) {
            auto streamPart = resultStreamQuery.ReadNext().ExtractValueSync();

            if (!streamPart.IsSuccess()) {
                eos = true;
                if (!streamPart.EOS()) {
                    return streamPart;
                }
                continue;
            }

>             // It is possible to duplicate lines in the output stream due to an external retryer.

Reviewer: ```suggestion
            // It is possible for lines to be duplicated in the output stream due to an external retrier
```

            if (streamPart.HasResultSet()) {
                auto rs = streamPart.ExtractResultSet();
                TResultSetParser parser(rs);
                while (parser.TryNextRow()) {
                    Cout << "Season"
                            << ", SeriesId: " << parser.ColumnParser("series_id").GetOptionalUint64()
                            << ", SeasonId: " << parser.ColumnParser("season_id").GetOptionalUint64()
                            << ", Title: " << parser.ColumnParser("title").GetOptionalUtf8()
                            << ", Air date: " << parser.ColumnParser("first_aired").GetOptionalDate()->FormatLocalTime("%Y-%m-%d")
                            << Endl;
                }
            }
        }
        return TStatus(EStatus::SUCCESS, NYql::TIssues());
    }));

}
```

> The given code snippet prints the following text to the console at startup (there may be duplicate lines in the output stream due to an external RetryQuerySync):

Reviewer: ```suggestion
The given code snippet prints the following text to the console at startup (there may be duplicate lines in the output stream due to an external `RetryQuerySync`):
```

```bash
> SelectWithParams:
Season, title: Season 3, series title: Silicon Valley
```

## Stream queries {#stream-query}

Making a stream query that results in a data stream. Streaming lets you read an unlimited number of rows and amount of data.

**WARNING**: Do not use without RetryQuery.

```c++
void StreamQuerySelect(TQueryClient client, const TString& path) {
    std::vector <TResultSet> resultSets;
    // WARNING: Do not use without RetryQuery!!!
    ThrowOnError(client.RetryQuerySync([path, &resultSets](TQueryClient client) -> TStatus {
        resultSets.clear();
        auto query = Sprintf(R"(
            --!syntax_v1
            PRAGMA TablePathPrefix("%s");

            DECLARE $series AS List<UInt64>;

            SELECT series_id, season_id, title, CAST(CAST(first_aired AS Date) AS String) AS first_aired
            FROM seasons
            WHERE series_id IN $series
            ORDER BY season_id;
        )", path.c_str());

        auto parameters = TParamsBuilder()
            .AddParam("$series")
            .BeginList()
                .AddListItem().Uint64(1)

>                 .AddListItem().Uint64(10)

Author: Did I understand correctly that you want the example to look like this?
``` cpp
std::vector<uint64_t> values = {1, 10};
auto parameters = TParamsBuilder()
        .AddParam("$series")
        .BeginList()
            .AddListItem().Uint64(values[0])
            .AddListItem().Uint64(values[1])
        .EndList().Build()
                    .Build();
```

>                 .AddListItem().Uint64(10)

Author: Adding values from std::vector is not implemented

>                 .AddListItem().Uint64(10)

Reviewer: You could still iterate over a vector and add values one by one?

            .EndList().Build()
            .Build();

        // Executes stream query
        auto resultStreamQuery = client.StreamExecuteQuery(query, TTxControl::NoTx(), parameters).GetValueSync();

        if (!resultStreamQuery.IsSuccess()) {
            return resultStreamQuery;
        }

        bool eos = false;

        while (!eos) {
            auto streamPart = resultStreamQuery.ReadNext().ExtractValueSync();

            if (!streamPart.IsSuccess()) {
                eos = true;
                if (!streamPart.EOS()) {
                    return streamPart;
                }
                continue;
            }

            if (streamPart.HasResultSet()) {
                auto rs = streamPart.ExtractResultSet();
                resultSets.push_back(rs);
            }
        }
        return TStatus(EStatus::SUCCESS, NYql::TIssues());
    }));
    
    Cout << "> StreamQuery:" << Endl;

>     for (auto rs : resultSets) {

Reviewer 2: It isn't correct because all execution is wrapped in RetryQuery, due to this output may be duplicated

            --!syntax_v1
            PRAGMA TablePathPrefix("%s");

            DECLARE $seriesId AS Uint64;
            DECLARE $seasonId AS Uint64;

            SELECT first_aired AS from_date FROM seasons
            WHERE series_id = $seriesId AND season_id = $seasonId;
        )", path.c_str());

        auto params1 = TParamsBuilder()
            .AddParam("$seriesId")
                .Uint64(seriesId)
                .Build()
            .AddParam("$seasonId")
                .Uint64(seasonId)
                .Build()
            .Build();

        // Execute first query to get the required values to the client.
        // Transaction control settings don't set CommitTx flag to keep transaction active
        // after query execution.
        auto result = session.ExecuteQuery(
            query1,
            TTxControl::BeginTx(TTxSettings::SerializableRW()),
            params1);

        auto resultValue = result.GetValueSync();

        if (!resultValue.IsSuccess()) {
            return resultValue;
        }
```

To continue working within the current transaction, you need to get the current `transaction id`:

```c++
        // Get active transaction id

>         auto tx = resultValue.GetTransaction();

Reviewer: As far as I understand, this doesn't actually obtain the transaction id. Something along the lines of `auto txId = resultValue.GetTransaction()->GetId();` would match the description better.

                .Build()
            .AddParam("$toDate")
                .Uint64(toDate.Days())
                .Build()
            .Build();

        // Execute second query.
        // Transaction control settings continues active transaction (tx) and
        // commits it at the end of second query execution.
        auto result2 = session.ExecuteQuery(
            query2,
            TTxControl::Tx(tx->GetId()).CommitTx(),
            params2).GetValueSync();
        
        if (!result2.IsSuccess()) {
            return result2;
        }
        resultSet = result2.GetResultSet(0);
        return result2;
    }));

    TResultSetParser parser(*resultSet);
    Cout << "> MultiStep:" << Endl;
    while (parser.TryNextRow()) {
        auto airDate = TInstant::Days(*parser.ColumnParser("air_date").GetOptionalUint64());

        Cout << "Episode " << parser.ColumnParser("episode_id").GetOptionalUint64()
            << ", Season: " << parser.ColumnParser("season_id").GetOptionalUint64()
            << ", Title: " << parser.ColumnParser("title").GetOptionalUtf8()
            << ", Air date: " << airDate.FormatLocalTime("%a %b %d, %Y")
            << Endl;
    }

> }

Reviewer: Among these multiple snippets I kinda lost track of where this lambda that is being retried ends; let's add a comment indicating it? Probably somewhere around here?

Code snippet for `BeginTransaction` and `tx.Commit()` calls:

```c++
void ExplicitTcl(TQueryClient client, const TString& path) {

>     // Show usage of explicit Begin/Commit transaction control calls.
>     // In most cases it's better to use transaction control settings in ExecuteDataQuery calls instead
>     // to avoid additional hops to YDB cluster and allow more efficient execution of queries.

Reviewer: ```suggestion
// Demonstrate the use of explicit Begin and Commit transaction control calls.
// In most cases, it's preferable to use transaction control settings within ExecuteDataQuery calls instead, 
// as this avoids additional hops to the YDB cluster and allows for more efficient query execution.
```

    ThrowOnError(client.RetryQuerySync([&path](TQueryClient client) -> TStatus {
        auto airDate = TInstant::Now();

>         auto session = client.GetSession().GetValueSync().GetSession();

Reviewer: Why `GetSession()` is called twice? The first one probably returns a promise; hence, `GetValueSync()` is needed, but what does the second `GetSession()` do?

Reviewer 2: GetValueSync() returns TCreateSessionResult that contains session which can be obtained from GetSession method
